{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "from opengsync_db import categories, DBHandler\n",
    "from IPython.display import display\n",
    "import pandas as pd\n",
    "from dotenv import load_dotenv\n",
    "import os\n",
    "load_dotenv()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOG: Connected to DB 'postgresql+psycopg://localhost:5432/opengsync_db'\n"
     ]
    }
   ],
   "source": [
    "db = DBHandler()\n",
    "db.connect(\n",
    "    user=os.environ[\"POSTGRES_USER\"],\n",
    "    password=os.environ[\"POSTGRES_PASSWORD\"],\n",
    "    host=os.environ[\"POSTGRES_SERVER_IP\"],\n",
    "    port=os.environ[\"POSTGRES_PORT\"],\n",
    "    db=os.environ[\"POSTGRES_DB\"],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 0. Categories (Enums)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(101, 'Poly-A RNA-Seq', 2697049)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# These are used to store statuses or types, e.g. LibraryType, ExperimentStatus, GenomeRef\n",
    "# Each enum has a unique id inside it's own enum-class\n",
    "# these are defined in services/opengsync-app/opengsync-db/opengsync_db/categories/\n",
    "categories.LibraryType.POLY_A_RNA_SEQ.id, categories.LibraryType.POLY_A_RNA_SEQ.name, categories.GenomeRef.COVID.organism_tax_id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Get All Libraries from Sequencing Run (Before demultiplexing flowcell)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Experiment(id=18, name=BSF_1764, num_lanes=2)\n",
      "S1 [2]\n",
      "User(id=3, email=csuete@cemm.at)\n",
      "2\n",
      "Archived [10]\n"
     ]
    }
   ],
   "source": [
    "experiment = db.get_experiment(name=\"BSF_1764\")\n",
    "print(experiment)\n",
    "print(experiment.flowcell_type)\n",
    "print(experiment.operator)\n",
    "print(experiment.num_lanes)\n",
    "print(experiment.status)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lane</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>library_name</th>\n",
       "      <th>library_type</th>\n",
       "      <th>reference</th>\n",
       "      <th>seq_request_id</th>\n",
       "      <th>sequence_i7</th>\n",
       "      <th>sequence_i5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1</td>\n",
       "      <td>B2_HC_D_CD4</td>\n",
       "      <td>B2_HC_D_CD4_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>ACTCGCTA</td>\n",
       "      <td>TCTACTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>1</td>\n",
       "      <td>B4_RA_R_Bcell</td>\n",
       "      <td>B4_RA_R_Bcell_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>GCTACGCT</td>\n",
       "      <td>ACTCTAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>1</td>\n",
       "      <td>B2_RA_P_Monocytes</td>\n",
       "      <td>B2_RA_P_Monocytes_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>TAGCGCTC</td>\n",
       "      <td>CTCCTTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_RA_G_CD8</td>\n",
       "      <td>B1_RA_G_CD8_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>ACTCGCTA</td>\n",
       "      <td>AGAGGATA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>2</td>\n",
       "      <td>15_5</td>\n",
       "      <td>15_5_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>AGGCAGAA</td>\n",
       "      <td>GTAAGGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>1</td>\n",
       "      <td>B2_RA_R_CD8</td>\n",
       "      <td>B2_RA_R_CD8_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>CCTAAGAC</td>\n",
       "      <td>CTCCTTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>1</td>\n",
       "      <td>B5_RA_R_Bcell</td>\n",
       "      <td>B5_RA_R_Bcell_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>TAGCGCTC</td>\n",
       "      <td>CTAGTCGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>1</td>\n",
       "      <td>B3_RA_R_CD8</td>\n",
       "      <td>B3_RA_R_CD8_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>AAGAGGCA</td>\n",
       "      <td>CTAGTCGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>1</td>\n",
       "      <td>B5_RA_R_CD4</td>\n",
       "      <td>B5_RA_R_CD4_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>ACTGAGCG</td>\n",
       "      <td>CTAGTCGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>1</td>\n",
       "      <td>B2_RA_P_CD8</td>\n",
       "      <td>B2_RA_P_CD8_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>ATGCGCAG</td>\n",
       "      <td>CTCCTTAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     lane        sample_name            library_name     library_type  \\\n",
       "25      1        B2_HC_D_CD4        B2_HC_D_CD4_ATAC   ATAC-Seq [108]   \n",
       "91      1      B4_RA_R_Bcell      B4_RA_R_Bcell_ATAC   ATAC-Seq [108]   \n",
       "43      1  B2_RA_P_Monocytes  B2_RA_P_Monocytes_ATAC   ATAC-Seq [108]   \n",
       "14      1        B1_RA_G_CD8        B1_RA_G_CD8_ATAC   ATAC-Seq [108]   \n",
       "144     2               15_5           15_5_SMARTSEQ  SMART-Seq [102]   \n",
       "45      1        B2_RA_R_CD8        B2_RA_R_CD8_ATAC   ATAC-Seq [108]   \n",
       "115     1      B5_RA_R_Bcell      B5_RA_R_Bcell_ATAC   ATAC-Seq [108]   \n",
       "69      1        B3_RA_R_CD8        B3_RA_R_CD8_ATAC   ATAC-Seq [108]   \n",
       "116     1        B5_RA_R_CD4        B5_RA_R_CD4_ATAC   ATAC-Seq [108]   \n",
       "42      1        B2_RA_P_CD8        B2_RA_P_CD8_ATAC   ATAC-Seq [108]   \n",
       "\n",
       "              reference  seq_request_id sequence_i7 sequence_i5  \n",
       "25   Human (GRCh38) [1]              60    ACTCGCTA    TCTACTCT  \n",
       "91   Human (GRCh38) [1]              60    GCTACGCT    ACTCTAGG  \n",
       "43   Human (GRCh38) [1]              60    TAGCGCTC    CTCCTTAC  \n",
       "14   Human (GRCh38) [1]              60    ACTCGCTA    AGAGGATA  \n",
       "144  Human (GRCh38) [1]              28    AGGCAGAA    GTAAGGAG  \n",
       "45   Human (GRCh38) [1]              60    CCTAAGAC    CTCCTTAC  \n",
       "115  Human (GRCh38) [1]              60    TAGCGCTC    CTAGTCGA  \n",
       "69   Human (GRCh38) [1]              60    AAGAGGCA    CTAGTCGA  \n",
       "116  Human (GRCh38) [1]              60    ACTGAGCG    CTAGTCGA  \n",
       "42   Human (GRCh38) [1]              60    ATGCGCAG    CTCCTTAC  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This dataframe should be sufficient to demultiplex a flowcell\n",
    "experiment_libraries = db.get_flowcell_df(experiment_id=experiment.id)\n",
    "experiment_libraries.sample(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.1 Grouping by Sequencing Request"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "seq_request_id: 28\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lane</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>library_name</th>\n",
       "      <th>library_type</th>\n",
       "      <th>reference</th>\n",
       "      <th>seq_request_id</th>\n",
       "      <th>sequence_i7</th>\n",
       "      <th>sequence_i5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>2</td>\n",
       "      <td>07_13</td>\n",
       "      <td>07_13_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>TAAGGCGA</td>\n",
       "      <td>CTCTCTAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>2</td>\n",
       "      <td>07_14</td>\n",
       "      <td>07_14_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>CGTACTAG</td>\n",
       "      <td>CTCTCTAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>2</td>\n",
       "      <td>07_15</td>\n",
       "      <td>07_15_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>AGGCAGAA</td>\n",
       "      <td>CTCTCTAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>2</td>\n",
       "      <td>07_16</td>\n",
       "      <td>07_16_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>TCCTGAGC</td>\n",
       "      <td>CTCTCTAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>2</td>\n",
       "      <td>13_1</td>\n",
       "      <td>13_1_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>GGACTCCT</td>\n",
       "      <td>CTCTCTAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>2</td>\n",
       "      <td>13_10</td>\n",
       "      <td>13_10_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>TAGGCATG</td>\n",
       "      <td>CTCTCTAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>2</td>\n",
       "      <td>13_11</td>\n",
       "      <td>13_11_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>CTCTCTAC</td>\n",
       "      <td>CTCTCTAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>2</td>\n",
       "      <td>13_12</td>\n",
       "      <td>13_12_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>CGAGGCTG</td>\n",
       "      <td>CTCTCTAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>2</td>\n",
       "      <td>13_2</td>\n",
       "      <td>13_2_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>AAGAGGCA</td>\n",
       "      <td>CTCTCTAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>2</td>\n",
       "      <td>13_3</td>\n",
       "      <td>13_3_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>GTAGAGGA</td>\n",
       "      <td>CTCTCTAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>2</td>\n",
       "      <td>13_4</td>\n",
       "      <td>13_4_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>GCTCATGA</td>\n",
       "      <td>CTCTCTAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>2</td>\n",
       "      <td>13_5</td>\n",
       "      <td>13_5_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>ATCTCAGG</td>\n",
       "      <td>CTCTCTAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>2</td>\n",
       "      <td>13_6</td>\n",
       "      <td>13_6_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>TAAGGCGA</td>\n",
       "      <td>TATCCTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>2</td>\n",
       "      <td>13_7</td>\n",
       "      <td>13_7_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>CGTACTAG</td>\n",
       "      <td>TATCCTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>2</td>\n",
       "      <td>13_8</td>\n",
       "      <td>13_8_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>AGGCAGAA</td>\n",
       "      <td>TATCCTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>2</td>\n",
       "      <td>13_9</td>\n",
       "      <td>13_9_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>TCCTGAGC</td>\n",
       "      <td>TATCCTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>2</td>\n",
       "      <td>14_1</td>\n",
       "      <td>14_1_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>GGACTCCT</td>\n",
       "      <td>TATCCTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>2</td>\n",
       "      <td>14_2</td>\n",
       "      <td>14_2_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>TAGGCATG</td>\n",
       "      <td>TATCCTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>2</td>\n",
       "      <td>14_3</td>\n",
       "      <td>14_3_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>CTCTCTAC</td>\n",
       "      <td>TATCCTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>2</td>\n",
       "      <td>14_4</td>\n",
       "      <td>14_4_SMARTSEQ</td>\n",
       "      <td>SMART-Seq [102]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>28</td>\n",
       "      <td>CGAGGCTG</td>\n",
       "      <td>TATCCTCT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     lane sample_name    library_name     library_type           reference  \\\n",
       "119     2       07_13  07_13_SMARTSEQ  SMART-Seq [102]  Human (GRCh38) [1]   \n",
       "120     2       07_14  07_14_SMARTSEQ  SMART-Seq [102]  Human (GRCh38) [1]   \n",
       "121     2       07_15  07_15_SMARTSEQ  SMART-Seq [102]  Human (GRCh38) [1]   \n",
       "122     2       07_16  07_16_SMARTSEQ  SMART-Seq [102]  Human (GRCh38) [1]   \n",
       "123     2        13_1   13_1_SMARTSEQ  SMART-Seq [102]  Human (GRCh38) [1]   \n",
       "124     2       13_10  13_10_SMARTSEQ  SMART-Seq [102]  Human (GRCh38) [1]   \n",
       "125     2       13_11  13_11_SMARTSEQ  SMART-Seq [102]  Human (GRCh38) [1]   \n",
       "126     2       13_12  13_12_SMARTSEQ  SMART-Seq [102]  Human (GRCh38) [1]   \n",
       "127     2        13_2   13_2_SMARTSEQ  SMART-Seq [102]  Human (GRCh38) [1]   \n",
       "128     2        13_3   13_3_SMARTSEQ  SMART-Seq [102]  Human (GRCh38) [1]   \n",
       "129     2        13_4   13_4_SMARTSEQ  SMART-Seq [102]  Human (GRCh38) [1]   \n",
       "130     2        13_5   13_5_SMARTSEQ  SMART-Seq [102]  Human (GRCh38) [1]   \n",
       "131     2        13_6   13_6_SMARTSEQ  SMART-Seq [102]  Human (GRCh38) [1]   \n",
       "132     2        13_7   13_7_SMARTSEQ  SMART-Seq [102]  Human (GRCh38) [1]   \n",
       "133     2        13_8   13_8_SMARTSEQ  SMART-Seq [102]  Human (GRCh38) [1]   \n",
       "134     2        13_9   13_9_SMARTSEQ  SMART-Seq [102]  Human (GRCh38) [1]   \n",
       "135     2        14_1   14_1_SMARTSEQ  SMART-Seq [102]  Human (GRCh38) [1]   \n",
       "136     2        14_2   14_2_SMARTSEQ  SMART-Seq [102]  Human (GRCh38) [1]   \n",
       "137     2        14_3   14_3_SMARTSEQ  SMART-Seq [102]  Human (GRCh38) [1]   \n",
       "138     2        14_4   14_4_SMARTSEQ  SMART-Seq [102]  Human (GRCh38) [1]   \n",
       "\n",
       "     seq_request_id sequence_i7 sequence_i5  \n",
       "119              28    TAAGGCGA    CTCTCTAT  \n",
       "120              28    CGTACTAG    CTCTCTAT  \n",
       "121              28    AGGCAGAA    CTCTCTAT  \n",
       "122              28    TCCTGAGC    CTCTCTAT  \n",
       "123              28    GGACTCCT    CTCTCTAT  \n",
       "124              28    TAGGCATG    CTCTCTAT  \n",
       "125              28    CTCTCTAC    CTCTCTAT  \n",
       "126              28    CGAGGCTG    CTCTCTAT  \n",
       "127              28    AAGAGGCA    CTCTCTAT  \n",
       "128              28    GTAGAGGA    CTCTCTAT  \n",
       "129              28    GCTCATGA    CTCTCTAT  \n",
       "130              28    ATCTCAGG    CTCTCTAT  \n",
       "131              28    TAAGGCGA    TATCCTCT  \n",
       "132              28    CGTACTAG    TATCCTCT  \n",
       "133              28    AGGCAGAA    TATCCTCT  \n",
       "134              28    TCCTGAGC    TATCCTCT  \n",
       "135              28    GGACTCCT    TATCCTCT  \n",
       "136              28    TAGGCATG    TATCCTCT  \n",
       "137              28    CTCTCTAC    TATCCTCT  \n",
       "138              28    CGAGGCTG    TATCCTCT  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "...\n",
      "\n",
      "seq_request_id: 60\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lane</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>library_name</th>\n",
       "      <th>library_type</th>\n",
       "      <th>reference</th>\n",
       "      <th>seq_request_id</th>\n",
       "      <th>sequence_i7</th>\n",
       "      <th>sequence_i5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_HC_D_Bcell</td>\n",
       "      <td>B1_HC_D_Bcell_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>GCTCATGA</td>\n",
       "      <td>ATAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_HC_D_CD4</td>\n",
       "      <td>B1_HC_D_CD4_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>ATCTCAGG</td>\n",
       "      <td>ATAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_HC_D_CD8</td>\n",
       "      <td>B1_HC_D_CD8_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>ACTCGCTA</td>\n",
       "      <td>ATAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_HC_D_Monocytes</td>\n",
       "      <td>B1_HC_D_Monocytes_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>GGAGCTAC</td>\n",
       "      <td>ATAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_RA_A_Bcell</td>\n",
       "      <td>B1_RA_A_Bcell_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>GCGTAGTA</td>\n",
       "      <td>ATAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_RA_A_CD4</td>\n",
       "      <td>B1_RA_A_CD4_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>CGGAGCCT</td>\n",
       "      <td>ATAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_RA_A_CD8</td>\n",
       "      <td>B1_RA_A_CD8_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>TACGCTGC</td>\n",
       "      <td>ATAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_RA_A_Monocytes</td>\n",
       "      <td>B1_RA_A_Monocytes_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>ATGCGCAG</td>\n",
       "      <td>ATAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_RA_D_Bcell</td>\n",
       "      <td>B1_RA_D_Bcell_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>TAGCGCTC</td>\n",
       "      <td>ATAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_RA_D_CD4</td>\n",
       "      <td>B1_RA_D_CD4_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>ACTGAGCG</td>\n",
       "      <td>ATAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_RA_D_CD8</td>\n",
       "      <td>B1_RA_D_CD8_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>CCTAAGAC</td>\n",
       "      <td>ATAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_RA_D_Monocytes</td>\n",
       "      <td>B1_RA_D_Monocytes_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>CGATCAGT</td>\n",
       "      <td>ATAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_RA_G_Bcell</td>\n",
       "      <td>B1_RA_G_Bcell_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>GCTCATGA</td>\n",
       "      <td>AGAGGATA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_RA_G_CD4</td>\n",
       "      <td>B1_RA_G_CD4_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>ATCTCAGG</td>\n",
       "      <td>AGAGGATA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_RA_G_CD8</td>\n",
       "      <td>B1_RA_G_CD8_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>ACTCGCTA</td>\n",
       "      <td>AGAGGATA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_RA_G_Monocytes</td>\n",
       "      <td>B1_RA_G_Monocytes_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>GGAGCTAC</td>\n",
       "      <td>AGAGGATA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_RA_P_Bcell</td>\n",
       "      <td>B1_RA_P_Bcell_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>GCGTAGTA</td>\n",
       "      <td>AGAGGATA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_RA_P_CD4</td>\n",
       "      <td>B1_RA_P_CD4_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>CGGAGCCT</td>\n",
       "      <td>AGAGGATA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_RA_P_CD8</td>\n",
       "      <td>B1_RA_P_CD8_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>TACGCTGC</td>\n",
       "      <td>AGAGGATA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_RA_P_Monocytes</td>\n",
       "      <td>B1_RA_P_Monocytes_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>ATGCGCAG</td>\n",
       "      <td>AGAGGATA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    lane        sample_name            library_name    library_type  \\\n",
       "0      1      B1_HC_D_Bcell      B1_HC_D_Bcell_ATAC  ATAC-Seq [108]   \n",
       "1      1        B1_HC_D_CD4        B1_HC_D_CD4_ATAC  ATAC-Seq [108]   \n",
       "2      1        B1_HC_D_CD8        B1_HC_D_CD8_ATAC  ATAC-Seq [108]   \n",
       "3      1  B1_HC_D_Monocytes  B1_HC_D_Monocytes_ATAC  ATAC-Seq [108]   \n",
       "4      1      B1_RA_A_Bcell      B1_RA_A_Bcell_ATAC  ATAC-Seq [108]   \n",
       "5      1        B1_RA_A_CD4        B1_RA_A_CD4_ATAC  ATAC-Seq [108]   \n",
       "6      1        B1_RA_A_CD8        B1_RA_A_CD8_ATAC  ATAC-Seq [108]   \n",
       "7      1  B1_RA_A_Monocytes  B1_RA_A_Monocytes_ATAC  ATAC-Seq [108]   \n",
       "8      1      B1_RA_D_Bcell      B1_RA_D_Bcell_ATAC  ATAC-Seq [108]   \n",
       "9      1        B1_RA_D_CD4        B1_RA_D_CD4_ATAC  ATAC-Seq [108]   \n",
       "10     1        B1_RA_D_CD8        B1_RA_D_CD8_ATAC  ATAC-Seq [108]   \n",
       "11     1  B1_RA_D_Monocytes  B1_RA_D_Monocytes_ATAC  ATAC-Seq [108]   \n",
       "12     1      B1_RA_G_Bcell      B1_RA_G_Bcell_ATAC  ATAC-Seq [108]   \n",
       "13     1        B1_RA_G_CD4        B1_RA_G_CD4_ATAC  ATAC-Seq [108]   \n",
       "14     1        B1_RA_G_CD8        B1_RA_G_CD8_ATAC  ATAC-Seq [108]   \n",
       "15     1  B1_RA_G_Monocytes  B1_RA_G_Monocytes_ATAC  ATAC-Seq [108]   \n",
       "16     1      B1_RA_P_Bcell      B1_RA_P_Bcell_ATAC  ATAC-Seq [108]   \n",
       "17     1        B1_RA_P_CD4        B1_RA_P_CD4_ATAC  ATAC-Seq [108]   \n",
       "18     1        B1_RA_P_CD8        B1_RA_P_CD8_ATAC  ATAC-Seq [108]   \n",
       "19     1  B1_RA_P_Monocytes  B1_RA_P_Monocytes_ATAC  ATAC-Seq [108]   \n",
       "\n",
       "             reference  seq_request_id sequence_i7 sequence_i5  \n",
       "0   Human (GRCh38) [1]              60    GCTCATGA    ATAGAGAG  \n",
       "1   Human (GRCh38) [1]              60    ATCTCAGG    ATAGAGAG  \n",
       "2   Human (GRCh38) [1]              60    ACTCGCTA    ATAGAGAG  \n",
       "3   Human (GRCh38) [1]              60    GGAGCTAC    ATAGAGAG  \n",
       "4   Human (GRCh38) [1]              60    GCGTAGTA    ATAGAGAG  \n",
       "5   Human (GRCh38) [1]              60    CGGAGCCT    ATAGAGAG  \n",
       "6   Human (GRCh38) [1]              60    TACGCTGC    ATAGAGAG  \n",
       "7   Human (GRCh38) [1]              60    ATGCGCAG    ATAGAGAG  \n",
       "8   Human (GRCh38) [1]              60    TAGCGCTC    ATAGAGAG  \n",
       "9   Human (GRCh38) [1]              60    ACTGAGCG    ATAGAGAG  \n",
       "10  Human (GRCh38) [1]              60    CCTAAGAC    ATAGAGAG  \n",
       "11  Human (GRCh38) [1]              60    CGATCAGT    ATAGAGAG  \n",
       "12  Human (GRCh38) [1]              60    GCTCATGA    AGAGGATA  \n",
       "13  Human (GRCh38) [1]              60    ATCTCAGG    AGAGGATA  \n",
       "14  Human (GRCh38) [1]              60    ACTCGCTA    AGAGGATA  \n",
       "15  Human (GRCh38) [1]              60    GGAGCTAC    AGAGGATA  \n",
       "16  Human (GRCh38) [1]              60    GCGTAGTA    AGAGGATA  \n",
       "17  Human (GRCh38) [1]              60    CGGAGCCT    AGAGGATA  \n",
       "18  Human (GRCh38) [1]              60    TACGCTGC    AGAGGATA  \n",
       "19  Human (GRCh38) [1]              60    ATGCGCAG    AGAGGATA  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Demultiplexed fastq-files should be grouped by 'seq_request_id' as each Sequencing Request can have only one Requestor\n",
    "for (seq_request_id), df in experiment_libraries.groupby(\"seq_request_id\"):\n",
    "    print(f\"seq_request_id: {seq_request_id}\")\n",
    "    display(df.head(20))\n",
    "    print(\"...\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1.2 Filtering Library Type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lane</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>library_name</th>\n",
       "      <th>library_type</th>\n",
       "      <th>reference</th>\n",
       "      <th>seq_request_id</th>\n",
       "      <th>sequence_i7</th>\n",
       "      <th>sequence_i5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_HC_D_Bcell</td>\n",
       "      <td>B1_HC_D_Bcell_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>GCTCATGA</td>\n",
       "      <td>ATAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_HC_D_CD4</td>\n",
       "      <td>B1_HC_D_CD4_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>ATCTCAGG</td>\n",
       "      <td>ATAGAGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>B1_HC_D_CD8</td>\n",
       "      <td>B1_HC_D_CD8_ATAC</td>\n",
       "      <td>ATAC-Seq [108]</td>\n",
       "      <td>Human (GRCh38) [1]</td>\n",
       "      <td>60</td>\n",
       "      <td>ACTCGCTA</td>\n",
       "      <td>ATAGAGAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   lane    sample_name        library_name    library_type  \\\n",
       "0     1  B1_HC_D_Bcell  B1_HC_D_Bcell_ATAC  ATAC-Seq [108]   \n",
       "1     1    B1_HC_D_CD4    B1_HC_D_CD4_ATAC  ATAC-Seq [108]   \n",
       "2     1    B1_HC_D_CD8    B1_HC_D_CD8_ATAC  ATAC-Seq [108]   \n",
       "\n",
       "            reference  seq_request_id sequence_i7 sequence_i5  \n",
       "0  Human (GRCh38) [1]              60    GCTCATGA    ATAGAGAG  \n",
       "1  Human (GRCh38) [1]              60    ATCTCAGG    ATAGAGAG  \n",
       "2  Human (GRCh38) [1]              60    ACTCGCTA    ATAGAGAG  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiment_libraries[experiment_libraries[\"library_type\"] == categories.LibraryType.ATAC_SEQ].head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Downstream Projects (after demultiplexing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>sex</th>\n",
       "      <th>tissue</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>556</td>\n",
       "      <td>07_13</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>557</td>\n",
       "      <td>07_14</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>558</td>\n",
       "      <td>07_15</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>559</td>\n",
       "      <td>07_16</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>560</td>\n",
       "      <td>13_1</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>561</td>\n",
       "      <td>13_10</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>562</td>\n",
       "      <td>13_11</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>563</td>\n",
       "      <td>13_12</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>564</td>\n",
       "      <td>13_2</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>565</td>\n",
       "      <td>13_3</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>566</td>\n",
       "      <td>13_4</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>567</td>\n",
       "      <td>13_5</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>568</td>\n",
       "      <td>13_6</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>569</td>\n",
       "      <td>13_7</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>570</td>\n",
       "      <td>13_8</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>571</td>\n",
       "      <td>13_9</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>572</td>\n",
       "      <td>14_1</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>573</td>\n",
       "      <td>14_2</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>574</td>\n",
       "      <td>14_3</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>575</td>\n",
       "      <td>14_4</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>576</td>\n",
       "      <td>14_5</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>577</td>\n",
       "      <td>14_6</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>578</td>\n",
       "      <td>15_1</td>\n",
       "      <td>m</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>579</td>\n",
       "      <td>15_2</td>\n",
       "      <td>m</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>580</td>\n",
       "      <td>15_3</td>\n",
       "      <td>m</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>581</td>\n",
       "      <td>15_5</td>\n",
       "      <td>m</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>582</td>\n",
       "      <td>15_6</td>\n",
       "      <td>m</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>583</td>\n",
       "      <td>15_7</td>\n",
       "      <td>m</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>584</td>\n",
       "      <td>15_8</td>\n",
       "      <td>m</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>585</td>\n",
       "      <td>17_1</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>586</td>\n",
       "      <td>17_2</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>587</td>\n",
       "      <td>17_3</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>588</td>\n",
       "      <td>17_4</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>589</td>\n",
       "      <td>18_1</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>590</td>\n",
       "      <td>18_2</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>591</td>\n",
       "      <td>18_3</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>592</td>\n",
       "      <td>18_4</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample_id sample_name sex tissue    cell_type\n",
       "0         556       07_13   f   Skin  Fibroblasts\n",
       "1         557       07_14   f   Skin  Fibroblasts\n",
       "2         558       07_15   f   Skin  Fibroblasts\n",
       "3         559       07_16   f   Skin  Fibroblasts\n",
       "4         560        13_1   f   Skin  Fibroblasts\n",
       "5         561       13_10   f   Skin  Fibroblasts\n",
       "6         562       13_11   f   Skin  Fibroblasts\n",
       "7         563       13_12   f   Skin  Fibroblasts\n",
       "8         564        13_2   f   Skin  Fibroblasts\n",
       "9         565        13_3   f   Skin  Fibroblasts\n",
       "10        566        13_4   f   Skin  Fibroblasts\n",
       "11        567        13_5   f   Skin  Fibroblasts\n",
       "12        568        13_6   f   Skin  Fibroblasts\n",
       "13        569        13_7   f   Skin  Fibroblasts\n",
       "14        570        13_8   f   Skin  Fibroblasts\n",
       "15        571        13_9   f   Skin  Fibroblasts\n",
       "16        572        14_1   f   Skin  Fibroblasts\n",
       "17        573        14_2   f   Skin  Fibroblasts\n",
       "18        574        14_3   f   Skin  Fibroblasts\n",
       "19        575        14_4   f   Skin  Fibroblasts\n",
       "20        576        14_5   f   Skin  Fibroblasts\n",
       "21        577        14_6   f   Skin  Fibroblasts\n",
       "22        578        15_1   m   Skin  Fibroblasts\n",
       "23        579        15_2   m   Skin  Fibroblasts\n",
       "24        580        15_3   m   Skin  Fibroblasts\n",
       "25        581        15_5   m   Skin  Fibroblasts\n",
       "26        582        15_6   m   Skin  Fibroblasts\n",
       "27        583        15_7   m   Skin  Fibroblasts\n",
       "28        584        15_8   m   Skin  Fibroblasts\n",
       "29        585        17_1   f   Skin  Fibroblasts\n",
       "30        586        17_2   f   Skin  Fibroblasts\n",
       "31        587        17_3   f   Skin  Fibroblasts\n",
       "32        588        17_4   f   Skin  Fibroblasts\n",
       "33        589        18_1   f   Skin  Fibroblasts\n",
       "34        590        18_2   f   Skin  Fibroblasts\n",
       "35        591        18_3   f   Skin  Fibroblasts\n",
       "36        592        18_4   f   Skin  Fibroblasts"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db.get_project_samples_df(23)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>experiment_name</th>\n",
       "      <th>lane</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>pool_name</th>\n",
       "      <th>library_name</th>\n",
       "      <th>seq_request_id</th>\n",
       "      <th>sex</th>\n",
       "      <th>tissue</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>BSF_1764</td>\n",
       "      <td>2</td>\n",
       "      <td>14_3</td>\n",
       "      <td>574</td>\n",
       "      <td>20250422_SS2_DF_cyt_AK</td>\n",
       "      <td>14_3_SMARTSEQ</td>\n",
       "      <td>28</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>BSF_1764</td>\n",
       "      <td>2</td>\n",
       "      <td>15_5</td>\n",
       "      <td>581</td>\n",
       "      <td>20250422_SS2_DF_cyt_AK</td>\n",
       "      <td>15_5_SMARTSEQ</td>\n",
       "      <td>28</td>\n",
       "      <td>m</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>BSF_1764</td>\n",
       "      <td>2</td>\n",
       "      <td>15_6</td>\n",
       "      <td>582</td>\n",
       "      <td>20250422_SS2_DF_cyt_AK</td>\n",
       "      <td>15_6_SMARTSEQ</td>\n",
       "      <td>28</td>\n",
       "      <td>m</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BSF_1764</td>\n",
       "      <td>2</td>\n",
       "      <td>13_12</td>\n",
       "      <td>563</td>\n",
       "      <td>20250422_SS2_DF_cyt_AK</td>\n",
       "      <td>13_12_SMARTSEQ</td>\n",
       "      <td>28</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>BSF_1764</td>\n",
       "      <td>2</td>\n",
       "      <td>18_4</td>\n",
       "      <td>592</td>\n",
       "      <td>20250422_SS2_DF_cyt_AK</td>\n",
       "      <td>18_4_SMARTSEQ</td>\n",
       "      <td>28</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>BSF_1764</td>\n",
       "      <td>2</td>\n",
       "      <td>13_4</td>\n",
       "      <td>566</td>\n",
       "      <td>20250422_SS2_DF_cyt_AK</td>\n",
       "      <td>13_4_SMARTSEQ</td>\n",
       "      <td>28</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>BSF_1764</td>\n",
       "      <td>2</td>\n",
       "      <td>18_1</td>\n",
       "      <td>589</td>\n",
       "      <td>20250422_SS2_DF_cyt_AK</td>\n",
       "      <td>18_1_SMARTSEQ</td>\n",
       "      <td>28</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>BSF_1764</td>\n",
       "      <td>2</td>\n",
       "      <td>13_8</td>\n",
       "      <td>570</td>\n",
       "      <td>20250422_SS2_DF_cyt_AK</td>\n",
       "      <td>13_8_SMARTSEQ</td>\n",
       "      <td>28</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>BSF_1764</td>\n",
       "      <td>2</td>\n",
       "      <td>15_1</td>\n",
       "      <td>578</td>\n",
       "      <td>20250422_SS2_DF_cyt_AK</td>\n",
       "      <td>15_1_SMARTSEQ</td>\n",
       "      <td>28</td>\n",
       "      <td>m</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>BSF_1764</td>\n",
       "      <td>2</td>\n",
       "      <td>13_9</td>\n",
       "      <td>571</td>\n",
       "      <td>20250422_SS2_DF_cyt_AK</td>\n",
       "      <td>13_9_SMARTSEQ</td>\n",
       "      <td>28</td>\n",
       "      <td>f</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Fibroblasts</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   experiment_name  lane sample_name  sample_id               pool_name  \\\n",
       "18        BSF_1764     2        14_3        574  20250422_SS2_DF_cyt_AK   \n",
       "25        BSF_1764     2        15_5        581  20250422_SS2_DF_cyt_AK   \n",
       "26        BSF_1764     2        15_6        582  20250422_SS2_DF_cyt_AK   \n",
       "7         BSF_1764     2       13_12        563  20250422_SS2_DF_cyt_AK   \n",
       "36        BSF_1764     2        18_4        592  20250422_SS2_DF_cyt_AK   \n",
       "10        BSF_1764     2        13_4        566  20250422_SS2_DF_cyt_AK   \n",
       "33        BSF_1764     2        18_1        589  20250422_SS2_DF_cyt_AK   \n",
       "14        BSF_1764     2        13_8        570  20250422_SS2_DF_cyt_AK   \n",
       "22        BSF_1764     2        15_1        578  20250422_SS2_DF_cyt_AK   \n",
       "15        BSF_1764     2        13_9        571  20250422_SS2_DF_cyt_AK   \n",
       "\n",
       "      library_name  seq_request_id sex tissue    cell_type  \n",
       "18   14_3_SMARTSEQ              28   f   Skin  Fibroblasts  \n",
       "25   15_5_SMARTSEQ              28   m   Skin  Fibroblasts  \n",
       "26   15_6_SMARTSEQ              28   m   Skin  Fibroblasts  \n",
       "7   13_12_SMARTSEQ              28   f   Skin  Fibroblasts  \n",
       "36   18_4_SMARTSEQ              28   f   Skin  Fibroblasts  \n",
       "10   13_4_SMARTSEQ              28   f   Skin  Fibroblasts  \n",
       "33   18_1_SMARTSEQ              28   f   Skin  Fibroblasts  \n",
       "14   13_8_SMARTSEQ              28   f   Skin  Fibroblasts  \n",
       "22   15_1_SMARTSEQ              28   m   Skin  Fibroblasts  \n",
       "15   13_9_SMARTSEQ              28   f   Skin  Fibroblasts  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# this dataframe should be enough to find all fastq-files generated from demultiplexing step\n",
    "# projects can contain multiple requests, and libraries from multiple experiments\n",
    "project_libraries = db.get_project_libraries_df(23)\n",
    "project_libraries[[\"experiment_name\", \"lane\", \"sample_name\", \"sample_id\", \"pool_name\", \"library_name\", \"seq_request_id\", \"sex\", \"tissue\", \"cell_type\"]].sample(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_request_name</th>\n",
       "      <th>user_id</th>\n",
       "      <th>email</th>\n",
       "      <th>first_name</th>\n",
       "      <th>last_name</th>\n",
       "      <th>role_id</th>\n",
       "      <th>role</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Stary lab pooled sample SrcX Fibroblast Project</td>\n",
       "      <td>3</td>\n",
       "      <td>csuete@cemm.at</td>\n",
       "      <td>Carina</td>\n",
       "      <td>Suete</td>\n",
       "      <td>3</td>\n",
       "      <td>Technician [3]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  seq_request_name  user_id           email  \\\n",
       "0  Stary lab pooled sample SrcX Fibroblast Project        3  csuete@cemm.at   \n",
       "\n",
       "  first_name last_name  role_id            role  \n",
       "0     Carina     Suete        3  Technician [3]  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db.get_seq_requestor_df(28)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. More advanced usage using Class hierarchy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Contact(id=107, name=Aglaja Kopf, email=akopf@cemm.oeaw.ac.at, phone=123, address=None)\n",
      "[Comment(id=40, text=Request Accepted , timestamp=2025-05-06 15:50:59.792390+02:00, author_id=3)]\n",
      "Custom [0]\n"
     ]
    }
   ],
   "source": [
    "db.open_session()\n",
    "seq_request = db.get_seq_request(28)\n",
    "print(seq_request.contact_person)\n",
    "print(seq_request.comments)\n",
    "print(seq_request.data_delivery_mode)\n",
    "db.close_session()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "User(id=3, email=csuete@cemm.at)\n"
     ]
    }
   ],
   "source": [
    "db.open_session()\n",
    "project = db.get_project(23)\n",
    "print(project.owner)\n",
    "db.close_session()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "opengsync",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
